deps(quic): update quinn to 0.11 and libp2p-tls to 0.4.0

retrohacker opened the pull request:
## Description

This PR updates the following dependencies `quic`:

  * `quinn@0.10.2` -> `quinn@0.11.0`
  * `rustls@0.21.9` -> `rustls@0.23.4`

## Notes & open questions

The new rustls types now use borrowed references by default and require you to take ownership of them by cloning explicitly (gated behind the `alloc` feature).

I'm not very good with rust lifetimes and I don't think I'm approaching them correctly in this PR. The `<'static>` seems wrong for example.

While working on this PR, I noticed the `max_bytes`, `max_percentages` tests are flakey on my machine. They both fail for me intermittently on the master branch.

I have them commented out right now for local dev but they need to be added back in before merge.

I also believe this is a breaking change since we seem to expose rustls.

`quinn` has not yet released support for `rustls@0.23` or `ring@0.17`, I'm referencing it as a git dependency until their PR is merged.
 
## Change checklist

<!-- Please add a Changelog entry in the appropriate crates and bump the crate versions if needed. See <https://github.com/libp2p/rust-libp2p/blob/master/docs/release.md#development-between-releases>-->

- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] A changelog entry has been made in the appropriate crates

retrohacker committed:
wip: bump ring+rustls in transports/tls

retrohacker committed:
wip: bump ring+rustls in transports/quic

retrohacker committed:
wip: update identity

retrohacker committed:
self review

retrohacker committed:
wip: update websocket

retrohacker committed:
add default providers to websocket

jxs commented on identity/Cargo.toml:
```diff
@@ -30,7 +30,7 @@ void = { version = "1.0", optional = true }
 
 [target.'cfg(not(target_arch = "wasm32"))'.dependencies]
-ring = { version = "0.17.5", features = [ "alloc", "std"], default-features = false, optional = true }
+ring = { version = "0.17.8", features = [ "alloc", "std"], default-features = false, optional = true }
```
we can probably move `ring` to a workspace dependency right?

jxs commented on transports/quic/src/config.rs:
```diff
@@ -71,11 +74,13 @@ pub struct Config {
+        let client_tls_config = libp2p_tls::make_client_config(keypair, None).unwrap();
+        let client_quic_config = Arc::new(QuicClientConfig::try_from(client_tls_config).unwrap());
+        let server_tls_config = libp2p_tls::make_server_config(keypair).unwrap();
+        let server_quic_config = Arc::new(QuicServerConfig::try_from(server_tls_config).unwrap());
```
can we keep this as it was? We don't seem to require `client_tls_config` and `server_tls_config` here.

jxs commented on transports/quic/Cargo.toml:
```diff
@@ -18,14 +18,14 @@ libp2p-core = { workspace = true }
+quinn = { git = "https://github.com/quinn-rs/quinn", branch="rustls-0.22", default-features = false, features = ["rustls", "futures-io"] }
 rand = "0.8.5"
-rustls = { version = "0.21.9", default-features = false }
+rustls = { version = "0.23.4", default-features = false, features = ["ring"] }
```
same as `ring` can we move `rustls` to a workspace dependency?

jxs commented on misc/memory-connection-limits/tests/max_bytes.rs:
```diff
@@ -71,6 +71,7 @@ fn max_bytes() {
 
     std::thread::sleep(Duration::from_millis(100)); // Memory stats are only updated every 100ms internally, ensure they are up-to-date when we try to exceed it.
 
+    /*
```
can you provide the output when these tests fail? So that we can try to understand what's making them flaky and fix it

jxs commented on transports/quic/src/config.rs:
```diff
@@ -108,8 +113,8 @@ pub(crate) struct QuinnConfig {
-            client_tls_config,
-            server_tls_config,
+            client_quic_config,
+            server_quic_config,
```
can we keep the previous names?

jxs commented on transports/quic/src/transport.rs:
```diff
@@ -472,7 +472,19 @@ impl<P: Provider> Listener<P> {
+                }
+            }
+        }
+        .boxed();
```
this should still return the new `Incoming`, and the result of `incoming.accept` should be handled in [`Listener::poll_next`](https://github.com/libp2p/rust-libp2p/blob/master/transports/quic/src/transport.rs#L585), in case `Incoming::accept` returns an error `TransportEvent::ListenerError` should be returned

jxs commented on transports/websocket/src/tls.rs:
```diff
@@ -35,32 +36,37 @@ impl fmt::Debug for Config {
 /// Private key, DER-encoded ASN.1 in either PKCS#8 or PKCS#1 format.
-#[derive(Clone)]
-pub struct PrivateKey(rustls::PrivateKey);
+pub struct PrivateKey<'a>(PrivateKeyDer<'a>);
```
can we use the owned versions of these new structs? I.e keep`pub struct PrivateKey(PrivateKeyDer<'static>);`

jxs reviewed (commented):
Thanks for getting this started! Left some comments

jxs committed:
Merge branch 'master' of github.com:libp2p/rust-libp2p into retrohacker/ring@0.17.8

jxs committed:
update quinn to published 0.11

jxs committed:
Merge branch 'master' of github.com:libp2p/rust-libp2p into retrohacker/ring@0.17.8

jxs committed:
deps(tls): upgrade rustls to 0.23

jxs committed:
Merge branch 'master' of github.com:libp2p/rust-libp2p into retrohacker/ring@0.17.8

jxs committed:
Merge branch 'update-rustls-tls' of github.com:libp2p/rust-libp2p into retrohacker/ring@0.17.8

dependabot[bot] committed:
chore(deps): bump golang.org/x/net from 0.21.0 to 0.23.0

Pull-Request: #5325.

jxs committed:
re-enable tests

jxs committed:
address remaining issues

Wiezzel committed:
feat(quic): allow setting the MTU discovery upper bound

The default MTU discovery upper bound is too high for some environments and results in permanent warnings. As I'd like to be able to configure this setting, adding a new method for the `Config`.

Pull-Request: #5386.

jxs committed:
return ListenerError on failed incoming

jxs committed:
Merge branch 'master' of github.com:libp2p/rust-libp2p into update-rustls-tls

jxs committed:
update CHANGELOG.md

jxs committed:
update versions

jxs committed:
increase write_after_peer_dropped_stream delay

jxs committed:
Merge branch 'update-rustls-tls' of github.com:libp2p/rust-libp2p into retrohacker/ring@0.17.8

jxs committed:
update Cargo.lock

jxs commented on transports/quic/tests/smoke.rs:
```diff
@@ -408,7 +408,7 @@ async fn write_after_peer_dropped_stream() {
     let (stream_a, mut stream_b) = build_streams::<quic::async_std::Provider>().await;
     drop(stream_a);
-    futures_timer::Delay::new(Duration::from_millis(1)).await;
+    futures_timer::Delay::new(Duration::from_millis(10)).await;
```
test has become flaky with 1ms, 10 seems to solve it

jxs committed:
Merge branch 'master' of github.com:libp2p/rust-libp2p into retrohacker/ring@0.17.8

jxs committed:
Merge branch 'master' of github.com:libp2p/rust-libp2p into retrohacker/ring@0.17.8

jxs committed:
update libp2p-tls version

jxs committed:
Merge branch 'master' of github.com:libp2p/rust-libp2p into retrohacker/ring@0.17.8

jxs committed:
update Cargo.lock

jxs committed:
rollback version to 0.10.3
